### PR TITLE
Final changes to support the serialization with metapointers

### PR DESCRIPTION
**Pull request by ftomassetti**
This is the third and final PR related to this. It contains a refactoring of JsonSerialization that is now based on the low-level serialization functionalities. I have also split JsonSerializationTest in multiple classes.

### CODE REVIEW

File: core/src/main/java/org/lionweb/lioncore/java/api/CombinedNodeResolver.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import org.lionweb.lioncore.java.model.Node;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This combines several NodeResolvers.
+ */
+public class CombinedNodeResolver implements NodeResolver {
```
Comment (enikao) [inline core/src/main/java/org/lionweb/lioncore/java/api/CombinedNodeResolver.java:12]: Combined or Composite?

File: core/src/main/java/org/lionweb/lioncore/java/serialization/JsonSerialization.java [hunk 2]
```diff
@@ -80,373 +81,158 @@ public PrimitiveValuesSerialization getPrimitiveValuesSerialization() {
...
+        serializedNode.getProperties().forEach(serializedPropertyValue ->{
+            Property property = concept.getPropertyByMetaPointer(serializedPropertyValue.getMetaPointer());
+            Objects.requireNonNull(property, "Property with metaPointer " + serializedPropertyValue.getMetaPointer() + " not found");
+            Object unserializedValue = primitiveValuesSerialization.unserialize(property.getType().getID(), serializedPropertyValue.getValue());
+            node.setPropertyValue(property, unserializedValue);
+        });
+        return node;
+    }
+
+    private void populateNode(SerializedNode serializedNode, NodeResolver nodeResolver) {
```
Comment (enikao) [inline core/src/main/java/org/lionweb/lioncore/java/serialization/JsonSerialization.java:207]: to my taste this method is a bit long, we could split it up

**Review by enikao — APPROVED** [review]

File: core/src/main/java/org/lionweb/lioncore/java/api/CombinedNodeResolver.java (hunk 1 above)
Comment (ftomassetti) [reply core/src/main/java/org/lionweb/lioncore/java/api/CombinedNodeResolver.java:12]: Renaming to Composite

File: core/src/main/java/org/lionweb/lioncore/java/serialization/JsonSerialization.java [hunk 3]
```diff
@@ -80,373 +81,158 @@ public PrimitiveValuesSerialization getPrimitiveValuesSerialization() {
...
+        Node node = nodeResolver.resolve(serializedNode.getID());
+        if (node == null) {
+            throw new IllegalStateException();
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/org/lionweb/lioncore/java/serialization/JsonSerialization.java:210]: Adding error message

File: core/src/main/java/org/lionweb/lioncore/java/serialization/JsonSerialization.java [hunk 4]
```diff
@@ -80,373 +81,158 @@ public PrimitiveValuesSerialization getPrimitiveValuesSerialization() {
...
-        jsonObject.add("properties", properties);
-
-        JsonObject children = new JsonObject();
-        node.getConcept().allContainments().forEach(containment -> {
-            JsonArray serializedValue = new JsonArray();
-            node.getChildren(containment).forEach(c -> serializedValue.add(c.getID()));
-            String containmentID = containment.getID();
-            if (containmentID == null) {
-                throw new IllegalStateException("Containment ID should not be null for " + containment);
+    public SerializedChunk serializeNodesToSerializationBlock(List<Node> nodes) {
```
[the remark this answers is not in the document]
Comment (ftomassetti) [reply core/src/main/java/org/lionweb/lioncore/java/serialization/JsonSerialization.java:95]: You are right. I have now extracted 4 methods out of this one

File: core/src/main/java/org/lionweb/lioncore/java/serialization/JsonSerialization.java (hunk 2 above)
Comment (ftomassetti) [reply core/src/main/java/org/lionweb/lioncore/java/serialization/JsonSerialization.java:222]: I have now extracted two methods out of this one

**Comment by ftomassetti** [comment]
Thank you @enikao for the review!

**Pull request merged**